### DISCUSSION TITLE
Xbox Controller firmware updated but still showing version 5.14 when I read that 5.17 is available.

### DISCUSSION THREAD

**Discussion by johnmrichey** [pybricks/support#1545] [General]
I updated the Xbox controller using the Xbox Accessories app and it completed the update and now says it's up to date, but when I check the FW version over USB it still shows as 5.14.  I think this is why I can't get the Xbox controller to connect to the latest beta of pybricks.  Any thoughts please.

**Comment by laurensvalk** [comment]
Yes, updating to 5.17 is required. My controller came with 5.14 out of the box and wouldn't work. Updating did the trick for me.

If updating still gives you 5.14, maybe you can try updating the Accessories App?

![5 17](https://github.com/pybricks/support/assets/12326241/9a9ee843-1721-41ee-9fed-cc9938636675)

---------------

I looked at the Accessories app just now and had some trouble using it. I was getting this message. But if I quickly clicked close and then immediately `...` I was able to get in the config menu and do the update. Based on an online search, Microsoft seems to be aware of this issue which should hopefully be fixed soon by updating the Accessories app.

![xboxacc](https://github.com/pybricks/support/assets/12326241/9f847fbd-0b0e-4c11-b893-a02e9c519a1c)

**Comment by johnmrichey** [comment]
In the Xbox Accy App it shows FW 5.20 now.
![FW_xbox](https://github.com/pybricks/support/assets/63120157/3ae29752-9e84-4757-8b76-26e5d865e7d2)
![FW_xbox_ACCY_APP](https://github.com/pybricks/support/assets/63120157/5a34b0e8-321b-4ae2-96d0-2d69c8765832)

**Reply by laurensvalk** [reply]
Interesting, that sounds like a very recent update! I'll have a look to see if it's available for me.

**Comment by johnmrichey** [comment]
But as you can see here USB Prober clearly shows it's still running 5.14.
<img width="799" alt="Screenshot 2024-03-14 at 00 54 32" src="https://github.com/pybricks/support/assets/63120157/7c539c87-7c60-4e76-9c31-0af28592f48a">

**Reply by laurensvalk** [reply]
Could this USB info possibly be outdated/cached information?

The app reporting 5.20 should mean that that's currently on the controller. So it sounds like the update was successful.

It's probably worth trying it with your LEGO hub now.

**Reply by johnmrichey** [reply]
USB Prober is a real time tool, it's not cached therefore it's still not updated and is running as reported by the tool.

I can run the USB Beagle on it and see as well to verify but I trust USB Prober and use it all day for my work so I trust the output.

**Reply by johnmrichey** [reply]
I tried pairing to the TechnicHub with audi_blocks.py and no luck.

**Reply by laurensvalk** [reply]
Understood, but the app seems to report that 5.20 is on your controller. It probably wouldn't hurt to try it with your LEGO hub now :)

**Reply by johnmrichey** [reply]
No luck.  Still no pairing.

**Reply by johnmrichey** [reply]
Is there simple program one could write to test just pairing of the controller and output the profile class please?

**Comment by johnmrichey** [comment]
So it seems despite the Xbox accy app on Windows is running it clearly does not update the controller.

**Comment by johnmrichey** [comment]
Earlier today when I attempted the update it showed 5.17 was the version, and then just now it showed this 5.20 was available, but the same problem and it fails to update.

**Comment by BertLindeman** [comment]
Strange app. Just now the xbox acc app reports:

![afbeelding](https://github.com/pybricks/support/assets/8142081/c06a5265-f5ec-4dba-9a0b-a2e1efb41156)

And it also stated "No update needed" So my xbox controller was already on 5.20.7.0?

**Reply by johnmrichey** [reply]
Yeah I get the same result despite knowing the controller is running 5.14.

**Comment by laurensvalk** [comment]
If you install the NRF Connect app on your phone and you scan/connect to the controller, you can also see the version.

What do you see? Mine reports 5.20 since the update just now. 

![Screenshot_20240314-091725.png](https://github.com/pybricks/support/assets/12326241/580eeee2-5277-45f8-a0d1-4ad0017cfb98)

**Reply by johnmrichey** [reply]
Fails to read the properties.
![Screenshot_20240314-013550](https://github.com/pybricks/support/assets/63120157/be6fcd2e-8d2d-47f9-835d-0d3011f8a9a8)

**Reply by laurensvalk** [reply]
You have to click on the downward arrow next to a property to read it.

**Reply by johnmrichey** [reply]
I tried doing that but still won't read the properties.

**Reply by laurensvalk** [reply]
Are you able to bond with the controller?

![Screenshot_20240314-095608.png](https://github.com/pybricks/support/assets/12326241/cfc8fb2d-5279-4bca-a189-0e199d1c12d8)

Then when it asks, pair and connect?

![Screenshot_20240314-095614.png](https://github.com/pybricks/support/assets/12326241/33e5a691-b5c1-4d53-bfe3-3759ea4d2d09)

**Reply by laurensvalk** [reply]
And remember all this will only work correctly when the controller is in **pairing mode**

When it's not, everything will **appear** to work, but not work at all. (this is what makes the controller a bit confusing).

**Reply by johnmrichey** [reply]
Connected and Bonded but still won't read those strings.
![Screenshot_20240314-021029](https://github.com/pybricks/support/assets/63120157/a2b42fb1-a1ff-4b22-a34d-2e49e2ebf08f)

**Reply by BertLindeman** [reply]
I had the same (sort-off). 
Stopped the app on the phone and did the connect and bond again and then the value was read. 
No idea why . . . 
(I was looking for the battery level)

**Comment by johnmrichey** [comment]
ok finally got nRF connect app to read the firmware value and sure enough it appears to be updated now to 5.20.
![Screenshot_20240314-081027](https://github.com/pybricks/support/assets/63120157/35bcdb6c-a35f-43d3-9f49-aa3a1d6ea231)

**Reply by johnmrichey** [reply]
although USB still reports 5.14.  however, I am still unable to get the TechnicHub to connect to the controller.

**Reply by laurensvalk** [reply]
Good to know the update worked!

Did you put the controller in pairing mode again when trying it with the Technic Hub?

Does the light on the controller turn solid white? Which program are you running?

**Reply by johnmrichey** [reply]
I can make a simple program with object controller connect but it seems when I try to run car it fails to connect.  I have tried with various motor and car configurations but still won't run with several audi car programs I have tried.  Perhaps you can help to write a quick program to say Press A and Motor on A will turn, just to verify please.

**Reply by johnmrichey** [reply]
the light on the controller does indeed turn solid white when it connects.

**Reply by BertLindeman** [reply]
Batteries?

**Reply by johnmrichey** [reply]
Batteries are good and when I run a simple program that uses Controller object it definitely connects to the controller.

**Reply by johnmrichey** [reply]
Say I want to write a simple program that when Button A is pressed on the controller it turns Motor on A.  How would that look please?

**Reply by BertLindeman** [reply]
Maybe naive. I had something a bit similar here.
Tried to connect the xbox controller to the technic hub. And failed.
Short reason: the controller tried to connect to my tablet where I worked on beta.pybricks.com app.
Switched off the tablet, and now I have a connection again.

Simple program: light on the hub goes green and if A pressed red:
![pybricks_program_2024-03-14T16_29_52](https://github.com/pybricks/support/assets/8142081/993569fa-7d18-4f85-9d30-a36ff597f805)

[EDIT]
Maybe easier: the python code:

```python
from pybricks.hubs import TechnicHub
from pybricks.iodevices import XboxController
from pybricks.parameters import Button, Color

# Set up all devices.
technic_hub = TechnicHub()
controller = XboxController()


# The main program starts here.
technic_hub.light.on(Color.WHITE)
while True:
    if Button.A in controller.buttons.pressed():
        technic_hub.light.on(Color.RED)
    else:
        technic_hub.light.on(Color.GREEN)

```

**Reply by johnmrichey** [reply]
success with lights, thanks. controller functioning as expected.  now let's see if we can get these motors to turn as it might be a firmware issue with the motors.  I have definitely used these motors on other projects, including the BuildPi HAT.

**Reply by johnmrichey** [reply]
so added a Motor on A to the if / else statement and Motor on A turns.  now the question is why don't either of audi car examples work for me.  I have tried the original one I got from pybricks blog via racing brick and also the modified one from @vanczakp posted here on the discussions.

**Reply by vanczakp** [reply]
did someone try my code btw? have not recieved any feedback.

**Reply by johnmrichey** [reply]
I was trying it and now that I got the controller successfully working but I could not get it to run.  just like I could not get any other car object examples to work.

**Comment by johnmrichey** [comment]
When I try to run @vanczakp program, it loads but fails to connect to the controller, and the Motor on D starts on run and never stops.

**Reply by johnmrichey** [reply]
This is similar to run the original program I downloaded.

from pybricks.iodevices import XboxController
from pybricks.parameters import Direction, Port
from pybricks.pupdevices import Motor
from pybricks.robotics import Car

# Set up all devices.
front = Motor(Port.A, Direction.CLOCKWISE)
rear = Motor(Port.B, Direction.CLOCKWISE)
steer = Motor(Port.D, Direction.CLOCKWISE)
car = Car(steer, [front, rear])
controller = XboxController()


# The main program starts here.
while True:
    car.drive_power(controller.triggers()[1] - controller.triggers()[0])
    car.steer(controller.joystick_left()[0])

**Reply by vanczakp** [reply]
strange. So the motor works. The program loads with 0 motor power/speed value so it either got some signal from the controller then failed or some bug. The code got incorrect formatting when I posted and an admin facelifted it. Try the two options below, those are the original and the code version of it
I will try later copying the code from this forum and let see if there was a copy error.
Try this block code version. I think you can run it without licence.
https://drive.google.com/file/d/1GL_gt6sHEq_SWK56nTBDlJT-dJBotWj7/view?usp=sharing
Or try the .py code version wich is the same.
https://drive.google.com/file/d/12WNhSPz22RQmpeVGfhbZ5neriArEM6it/view?usp=sharing

**Reply by johnmrichey** [reply]
Same issue with this program.  Once it runs the controller fails to connect and Motor on D just runs on it's own.  The controller never connects despite now that I have confirmed we know the controller works and I can control motors.  In fact I tested each motor on A with the simple program and they all function as expected.

**Reply by johnmrichey** [reply]
I tried another TechnicHub out of the Audi e-tron set and see it behaves the same.  This is the one with the four screws vs. the easy to remove battery cover.  But same thing.  Loaded beta firmware from scratch, ran your program and the motor on D just run and the controller never connects.

**Reply by vanczakp** [reply]
Then it must be the controller. I just bought one and did not bother to check if it is updated. The code must be good as it was used and tested and most of it comes from Powered up. Do you happen to have the LR13000 crane? I can create a simple code for that to see if that works (with top hub, wont bother with broadcasting etc)

**Reply by johnmrichey** [reply]
Controller is working fine with a simple program.  I am sure the code is good, it looks fantastic.  Perhaps it has something to do with Chrome on MacOS Sonoma?  Sorry I don't have the Liebherr crane.

**Reply by BertLindeman** [reply]
The motor on port.D is the STEER motor.
The car class starts with calibration of the steer angles. So turns one side until stall and then the other side.
The middle between those angles is the steer-straight angle.

Is that what you see? Motor D turning, but does not stall as there is nothing to stall against?

[EDIT]
The car is instantiated **before** the xbox controller.
My assessment is there is no problem with the xbox controller.

**Reply by laurensvalk** [reply]
Yep - what @BertLindeman says (thanks Bert!) The steer motor runs until it hits one end stop, then the other, in order to find the center.

If there aren't any stops, it isn't going to find the center and just keep going. 

So probably try checking your build and the cables/ports.

But this does keep coming up, so maybe to make this more intuitive we can raise an error message if it still hasn't found the center after 10 seconds or so.

**Reply by johnmrichey** [reply]
Program starts and Motor on D just runs and never stops and the controller does not pair.  See attached movie.

https://github.com/pybricks/support/assets/63120157/1bb9a88a-7402-4263-9774-b1d131974a35

**Reply by BertLindeman** [reply]
Oke, movie proof:


https://github.com/pybricks/support/assets/8142081/0de3bbe1-c4f9-452c-a05e-59bcae1f5115

[EDIT]
The XL motor on the right is on port-D is calibrated and after that the xbox controller connects

**Reply by vanczakp** [reply]
@johnmrichey here is your solution as @BertLindeman said: wrong ports :) 
Two drive motors on A (red, rear) and B (blue and front) and D (yellow) as steering

**Reply by laurensvalk** [reply]
We may be talking past each other :)

@johnmrichey - just add another pin in one of the remaining holes on that motor.

Anything that blocks the motor from going round forever will do.

The motor is rotating until it can't go any further. In your 'car', this never happens.

**Reply by johnmrichey** [reply]
Success!  This is great.  I am also running the program from @vanczakp.  Thanks so much for your patience and guidance.  Time to build the e-tron now.

https://github.com/pybricks/support/assets/63120157/4511b89f-52ef-4eb5-83ae-a5b559d8baeb

**Reply by vanczakp** [reply]
@johnmrichey enjoy :)

**Comment by BertLindeman** [comment]
> success with lights, thanks. controller functioning as expected. now let's see if we can get these motors to turn as it might be a firmware issue with the motors. I have definitely used these motors on other projects, including the BuildPi HAT.

I assume you can adapt the "lights" program by adding a motor? (yoking . . ) (I see you already did)

Nasty that the Technic hub disconnects from the PC. If you have an android table at hand, you might try that with argument from #1462

```python
xbox = XboxController(stay_connected=True)
```
